Rewrite oracles

hushaohan opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add it in the CHANGELOG file under Unreleased section.
⚠️ If your pull request fixes an open issue, please link to the issue.

✅ I have added the tests to cover my changes.
✅ I have updated the documentation accordingly.
✅ I have read the CONTRIBUTING document.
-->

### Summary

rewrite oracle constructions and oracle-oriented algorithms.

### Details and comments

- `Oracle` subclasses are no longer characterized by which algorithm(s) they are attached to, but by how the oracle's circuit is to be constructed.

- For `Grover`, the old `SAT` oracle is replaced by `DimacsOracle`.

- For optimization on the newly added `TruthTableOracle`, implemented Quine-McCluskey algorithm for computing all the prime implicants, DLX is then used to find an exact cover.

- For optimization on other logic expression-oriented oracles, Berkeley's `espresso` as provided via `pyeda` is used for minimizing the expressions.

- `Bernstein-Varizani`'s and `Simon`'s oracle constructions no longer use any information from classically obtained ground-truth. In particular, they now rely on single- and multi-output `ESOP` to build oracle circuits that truthfully reflect valid input bitmaps, just like `Deutsch-Josza`.

- future work includes:
    - add additional optimizations to logic-based oracles
    - add oracle construction from user supplied circuit

- todo: update docs

pistoia commented:
Thanks, @hushaohan -- Can you please add the relevant documentation to the Qiskit repository?  Please do not forget to update the documentation for the existing `SAT` oracle for `Grover`, nor renamed `CNFOracle`.

hushaohan commented:
> Thanks, @hushaohan -- Can you please add the relevant documentation to the Qiskit repository? Please do not forget to update the documentation for the existing `SAT` oracle for `Grover`, now renamed `CNFOracle`.

Sure, will do

hushaohan commented:
> Thanks, @hushaohan -- Can you please add the relevant documentation to the Qiskit repository? Please do not forget to update the documentation for the existing `SAT` oracle for `Grover`, now renamed `CNFOracle`.

updated with https://github.com/Qiskit/qiskit/pull/127
